### PR TITLE
Fix build pipeline

### PR DESCRIPTION
**Pull request by saqib-cliqz**
@spacifici @sharath-cliqz 

Please have a look if something is missing.

### CODE REVIEW

File: mozilla-release/Jenkinsfile.android [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/bin/env groovy
+
+node('android') {
```
Comment (sharath-cliqz) [inline mozilla-release/Jenkinsfile.android]: Can you please change this line to:

node('us-east-1 && ubuntu && docker && !gpu') {

**Comment by spacifici** [comment]
@alver-cliqz can you please review?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    apt-get update && \
+    apt-get upgrade -y && \
+    apt-get install -y autoconf2.13 build-essential ccache curl git lib32z1 \
+        libc6:i386 libncurses5:i386 libstdc++6:i386 mercurial openjdk-8-jdk \
+        python-dev python-pip python-setuptools unzip uuid \
+        wget xz-utils zip && \
+    apt-get clean -y && \
+    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* && \
+    mkdir -p /root/.mozbuild/proguard/lib && \
+    wget --output-document=/tmp/proguard.tgz --quiet "https://vorboss.dl.sourceforge.net/project/proguard/proguard/5.3/proguard5.3.3.tar.gz" && \
```
Comment (alver-cliqz) [inline Dockerfile]: This is external dependencies, which can die at any moment. Better to copy to our S3 bucket.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    rm -r sdktools.zip; \
+    (while (true); do echo y; sleep 2; done) | \
+      tools/bin/sdkmanager  "build-tools;25.0.3" "platforms;android-23" "platform-tools" "tools" "extras;google;m2repository" "extras;android;m2repository" "extras;google;google_play_services" "emulator";
+RUN cd /root && \
+    wget --output-document=nodejs.tgz --quiet \
+      'https://nodejs.org/dist/v8.9.3/node-v8.9.3-linux-x64.tar.xz' && \
+    tar xf nodejs.tgz && \
+    rm -f nodejs.tgz
+ENV PATH "/root/node-v8.9.3-linux-x64/bin:$PATH"
+RUN npm install -g yarn
```
Comment (alver-cliqz) [inline Dockerfile]: Why you don't want just to launch "./mach bootstrap" as prepare step? In other case you will need to fix this file for each major FF version

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                ac_add_options --enable-application=mobile/android
+                ac_add_options --target=i386-linux-android
+                # With the following Android SDK:
+                ac_add_options --with-android-sdk=\"/root/.mozbuild/android-sdk-linux\"
+                # Enable artifact building:
+                ac_add_options --enable-artifact-builds
+                # Write build artifacts to:
+                mk_add_options MOZ_OBJDIR=./objdir-frontend
+                " >> mozconfig
+            '''
```
Comment (alver-cliqz) [inline Jenkinsfile]: Better to have this in master branch as default configuration.

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                }
+        }
+    }
+
+    def args = "-u 0:0 -v ${pwd}/artifacts:/artifacts:rw"
+    docker.image("141047255820.dkr.ecr.us-east-1.amazonaws.com/${baseImageName}").inside(args) {
+        stage('Rust Setup') {
+            sh '''#!/bin/bash -l
+                curl https://sh.rustup.rs -sSf | sh -s -- -y -v
+            '''
```
Comment (alver-cliqz) [inline Jenkinsfile]: Please, do this in a docker container preparation. I think this is not a good idea to install dependencies in two different places.

File: mozilla-release/Makefile.in [hunk 6]
```diff
@@ -210,7 +210,8 @@ binaries::
 endif
 
 recurse_artifact:
-	$(topsrcdir)/mach --log-no-times artifact install
+	$(topsrcdir)/mach --log-no-times artifact install https://queue.taskcluster.net/v1/task/J-VPXVOPSqOzeAsiLDsTcQ/artifacts/public/build/target.apk
+	$(topsrcdir)/mach --log-no-times artifact install https://queue.taskcluster.net/v1/task/J-VPXVOPSqOzeAsiLDsTcQ/artifacts/public/build/target.common.tests.zip
```
Comment (alver-cliqz) [inline mozilla-release/Makefile.in]: Please, change link to:
$(topsrcdir)/mach --log-no-times artifact install https://repository.cliqz.com/dist/android/artifacts/$(GRE_MILESTONE)/target.apk
$(topsrcdir)/mach --log-no-times artifact install https://repository.cliqz.com/dist/android/artifacts/$(GRE_MILESTONE)/target.common.tests.zip

This is from https://github.com/cliqz-oss/cliqz-android/pull/3, so we can close it without merging.

File: mozilla-release/mobile/android/moz.configure [hunk 7]
```diff
@@ -129,8 +129,9 @@ def check_install_tracking(install_tracking,
             die('You must specify MOZ_NATIVE_DEVICES=1 when'
                 ' building with MOZ_INSTALL_TRACKING=1')
         if not adjust_sdk_keyfile:
-            die('You must specify --with-adjust-sdk-keyfile=/path/to/keyfile when'
-                ' building with MOZ_INSTALL_TRACKING=1')
+            # Using default if Adjust SDK keyfile is not specified
+            # See https://issues.adblockplus.org/ticket/5920
+            adjust_sdk_keyfile = "$topsrcdir/mobile/android/base/adjust-sdk-sandbox.token"
```
Comment (alver-cliqz) [inline mozilla-release/mobile/android/moz.configure]: This is will not need in version 60. Please, don't forget to remove

File: Dockerfile (hunk 3 above)
Comment (sharath-cliqz) [reply Dockerfile]: It was taking 20-30mins, sometimes even more, just to run the './mach bootstrap' every PR. Wouldn't it be better to have the dependencies installed in the Docker and save the Docker in the registry ?

File: Dockerfile (hunk 2 above)
Comment (sharath-cliqz) [reply Dockerfile]: True. I will copy it to S3.

**Comment by spacifici** [comment]
@alver-cliqz Should I create a combined pull request between this one and your own one?

File: Dockerfile (hunk 3 above)
Comment (alver-cliqz) [reply Dockerfile]: Yes, it take a lot on fresh image. You can save and use image between major FF updates

File: Dockerfile (hunk 2 above)
Comment (chrmod) [reply Dockerfile]: also proguard is in ubuntu repository, at least it is in 17.10

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    rm -rf /tmp/proguard*
+ENV SHELL /bin/bash
+ENV ANDROID_HOME /root/.mozbuild/android-sdk-linux
+RUN mkdir -p $ANDROID_HOME; \
+    cd $ANDROID_HOME; \
+    wget --output-document=sdktools.zip --quiet 'https://dl.google.com/android/repository/sdk-tools-linux-3859397.zip'; \
+    unzip sdktools.zip; \
+    rm -r sdktools.zip; \
+    (while (true); do echo y; sleep 2; done) | \
+      tools/bin/sdkmanager  "build-tools;25.0.3" "platforms;android-23" "platform-tools" "tools" "extras;google;m2repository" "extras;android;m2repository" "extras;google;google_play_services" "emulator";
```
Comment (chrmod) [inline Dockerfile]: cannot we use some pre made docker image for android?

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+RUN cd /root && \
+    wget --output-document=nodejs.tgz --quiet \
+      'https://nodejs.org/dist/v8.9.3/node-v8.9.3-linux-x64.tar.xz' && \
+    tar xf nodejs.tgz && \
+    rm -f nodejs.tgz
```
Comment (chrmod) [inline Dockerfile]: please use `nvm` and install it on user level

File: Dockerfile (hunk 3 above)
Comment (chrmod) [reply Dockerfile]: I've suggested Sharath to not use bootstrap as we often run into problems using it. 
Valid alternative IMO would be to use one of Dockerfiles used by Mozilla.
@sharath-cliqz did you take a look on docker files in mozilla repository?

File: Jenkinsfile [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    stage('Checkout') {
+        checkout scm
+    }
+    sh "`aws ecr get-login --region=us-east-1 --no-include-email`"
+    def dockerfileChecksum = sh(returnStdout: true, script: 'md5sum Dockerfile | cut -d" " -f1').trim()
+    def dockerTag = "${dockerfileChecksum}"
+    def baseImageName = "browser-f/android:${dockerTag}"
+    stage('Build docker image') {
+        docker.withRegistry('https://141047255820.dkr.ecr.us-east-1.amazonaws.com'){
+                try {
```
Comment (chrmod) [inline Jenkinsfile]: please use consistent indentation

File: Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                    image.pull()
+                } catch (e) {
+                    print e
+                    def baseImage  = docker.build(baseImageName, '--build-arg UID=`id -u` --build-arg GID=`id -g` .')
+                    baseImage.push dockerTag
+                }
+        }
+    }
+
+    def args = "-u 0:0 -v ${pwd}/artifacts:/artifacts:rw"
```
Comment (chrmod) [inline Jenkinsfile]: I would definitely discourage using root user in docker container. In this way all artifact will have root rights and you wont be able to work with them outside of docker image without a sudo.

File: Jenkinsfile (hunk 4 above)
Comment (chrmod) [reply Jenkinsfile]: also, take a look on mozconfig files provided by mozilla, maybe there is something important set there already

File: Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                # Enable artifact building:
+                ac_add_options --enable-artifact-builds
+                # Write build artifacts to:
+                mk_add_options MOZ_OBJDIR=./objdir-frontend
+                " >> mozconfig
+            '''
+        }
+        stage('Compile APK') {
+            sh '''#!/bin/bash -l
+                export SHELL=/bin/bash
```
Comment (chrmod) [inline Jenkinsfile]: why this is needed?

File: Jenkinsfile [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+                cd mozilla-release
+                ./mach build
+                ./mach package
```
Comment (chrmod) [inline Jenkinsfile]: please separate `package` into own stage

File: Dockerfile (hunk 3 above)
Comment (spacifici) [reply Dockerfile]: @chrmod those are specific to their TaskCluster (https://firefox-source-docs.mozilla.org/taskcluster/taskcluster/docker-images.html). Perhaps, we need to pre-build a Docker image and push it on our own registry.

File: Jenkinsfile (hunk 4 above)
Comment (spacifici) [reply Jenkinsfile]: Do not put the default configuration in `mozilla-release`, copy it over (i.e. put there a `mozconfig.default` and copy it to `mozconfig`.
The provided example mozconfigs are quite interesting to read.

Comment (alver-cliqz) [reply Jenkinsfile]: You can specify your mozconfig file with MOZCONFIG env variable. Please, check desktop browser: https://github.com/cliqz-oss/browser-f/blob/master/cliqz_env.sh#L67

File: Jenkinsfile (hunk 13 above)
Comment (saqib-cliqz) [reply Jenkinsfile]: @sharath-cliqz Can you please comment to the Questions by @chrmod regarding Jenkinsfile and Dockerfile

**Pull request merged**